File: 3. Programmers/13. 크레인 인형뽑기 게임/main.swift
//
//  main.swift
//  Algorithm
//
//  Created by 김동현 on 8/8/25.
//

import Foundation

func solution(_ board:[[Int]], _ moves:[Int]) -> Int {
    var count: Int = 0
    var bucket: [Int] = []
    var stacks: [[Int]] = Array(repeating: [], count: board.count)
    
    board.reversed().forEach {
        $0.enumerated().forEach {
            if $0.1 != 0 {
                stacks[$0.0].append($0.1)
            }
        }
    }
    
    moves.forEach {
        if let doll = stacks[$0-1].popLast() {
            if let last = bucket.last, last == doll {
                bucket.removeLast(1)
                count += 2
            } else {
                bucket.append(doll)
            }
        }
    }
    
    return count
}


let board: [[Int]] = [
    [0, 0, 0, 0, 0],
    [0, 0, 1, 0, 3],
    [0, 2, 5, 0, 1],
    [4, 2, 4, 4, 2],
    [3, 5, 1, 3, 1]
]

let moves: [Int] = [1, 5, 3, 5, 1, 2, 1, 4]
print(solution(board, moves))

/*
import Foundation

let size = 500
let board: [[Int]] = (0..<size).map { _ in
    (0..<size).map { _ in Int.random(in: 0...5) }
}

// MARK: - forEach 방식
func setupStacksWithForEach(_ board: [[Int]]) -> [[Int]] {
    var stacks = Array(repeating: [Int](), count: board.count)

    board.reversed().forEach {
        $0.enumerated().forEach {
            if $0.1 != 0 {
                stacks[$0.0].append($0.1)
            }
        }
    }

    return stacks
}

// MARK: - for-in 방식
func setupStacksWithForIn(_ board: [[Int]]) -> [[Int]] {
    var stacks = Array(repeating: [Int](), count: board.count)

    for row in board.reversed() {
        for (col, doll) in row.enumerated() {
            if doll != 0 {
                stacks[col].append(doll)
            }
        }
    }

    return stacks
}

// MARK: - 시간 측정 함수
func measureExecutionTime(name: String, block: () -> Void) {
    let start = DispatchTime.now()
    block()
    let end = DispatchTime.now()
    let nanoTime = end.uptimeNanoseconds - start.uptimeNanoseconds
    let timeInterval = Double(nanoTime) / 1_000_000
    print("\(name) 실행 시간: \(timeInterval)ms")
}

// MARK: - 실행
measureExecutionTime(name: "forEach") {
    _ = setupStacksWithForEach(board)
}

measureExecutionTime(name: "for-in") {
    _ = setupStacksWithForIn(board)
}
*/

File: 3. Programmers/13. 크레인 인형뽑기 게임/main.cpp
//
//  main.cpp
//  Algorithm
//
//  Created by 김동현 on 8/10/25.
//

#include <iostream>
#include <vector>
using namespace std;


int solution(vector<vector<int>> board, vector<int> moves) {
    
    int count = 0;
    int boardSize = static_cast<int>(board.size()); // ← 명시적 변환
    vector<int> bucket;
    vector<vector<int>> stacks(boardSize);
    
    // for (size_t i = boardSize; i-- > 0; ) {
    for (int i = boardSize - 1; i >= 0; i--) {
        for (int j=0; j<boardSize; j++) {
            int doll = board[i][j];
            if (doll != 0) {
                stacks[j].push_back(doll);
            }
        }
    }
    
    for (const auto& move : moves) {
        if (!stacks[move-1].empty()) {
            int doll = stacks[move-1].back();
            stacks[move-1].pop_back();
            
            if (bucket.empty()) {
                bucket.push_back(doll);
            } else {
                if (bucket.back() == doll) {
                    bucket.pop_back();
                    count += 2;
                } else {
                    bucket.push_back(doll);
                }
            }
        }
    }
    return count;
}

int main() {
    cout << solution( {
        {0, 0, 0, 0, 0},
        {0, 0, 1, 0, 3},
        {0, 2, 5, 0, 1},
        {4, 2, 4, 4, 2},
        {3, 5, 1, 3, 1}
    }, {1, 5, 3, 5, 1, 2, 1, 4} )
    << endl; // 4
    return 0;
}
